### PR TITLE
[SPARK-22562][Streaming] CachedKafkaConsumer unsafe eviction from cache

### PR DESCRIPTION
**Pull request by daroo**
## What changes were proposed in this pull request?
Fixes a problem when one thread wants to add a new consumer into fully packed cache and another one still uses an instance of cached consumer which is marked for eviction. In such cases underlying KafkaConsumer throws ConcurrentModificationException.

My solution is to always remove the eldest consumer from the cache, but sometimes delay calling close() method (in separate thread) until is no longer used (released) by KafkaRDDIterator

## How was this patch tested?
Simple unit test: fails on the current master branch and doesn't after the patch

### CODE REVIEW

**Comment by daroo** [comment]
Hey @koeninger ,
could you please have a look at my PR?

Thanx a lot,
Darek

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/CachedKafkaConsumer.scala [hunk 1]
```diff
@@ -155,11 +178,11 @@ object CachedKafkaConsumer extends Logging {
         logInfo(s"Cache miss for $k")
         logDebug(cache.keySet.toString)
         val c = new CachedKafkaConsumer[K, V](groupId, topic, partition, kafkaParams)
-        cache.put(k, c)
+        cache.put(k, c.acquireAndGet())
         c
       } else {
         // any given topicpartition should have a consistent key and value type
-        v.asInstanceOf[CachedKafkaConsumer[K, V]]
+        v.acquireAndGet().asInstanceOf[CachedKafkaConsumer[K, V]]
```
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/CachedKafkaConsumer.scala]: Shouldn't this method call be after the cast?

**Comment by koeninger** [comment]
My main comment is that if you have a situation where there's actually contention on the size of the cache, chances are things are going to be screwed up anyway due to consumers being recreated and losing any benefit of buffering.

Are you running into this issue in situations where the cache is appropriately sized given the number of topicpartitions and executors?

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/CachedKafkaConsumer.scala (hunk 1 above)
Comment (daroo) [reply external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/CachedKafkaConsumer.scala]: I can be, but actually it doesn't matter as the cache value has CachedKafkaConsumer[_, _] type

**Comment by daroo** [comment]
I see you point, but it's kind of difficult to size it properly when you use SubscribePattern (i.e. dynamic number of topics/partitions) consumer strategy

**Comment by koeninger** [comment]
Yeah, subscribepattern could definitely be an issue.

As far as unit testing, have you tried anything along the lines of setting
the cache size artificially low and then introducing new topicpartitions?

**Comment by daroo** [comment]
I've added a test. @koeninger is it something you had in mind?

**Comment by koeninger** [comment]
What are you actually asserting in that test and/or does it reliably fail
if run on the version of your code before the patch?

**Comment by daroo** [comment]
It fails on the current master branch and doesn't after the patch

**Comment by koeninger** [comment]
Seems reasonable.

**Comment by daroo** [comment]
Cool. Could you please authorize it for testing?

**Comment by koeninger** [comment]
ok to test

**Comment by daroo** [comment]
It seems that your "magic spell" didn't work. No build was triggered

**Comment by koeninger** [comment]
You'll need to get a commiter's attention to merge it anyway

On Nov 23, 2017 01:48, "Daroo" <notifications@github.com> wrote:

It seems that your "magic spell" didn't work. No build was triggered

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/19789#issuecomment-346548898>, or mute
the thread

**Comment by daroo** [comment]
Hey Spark commiters @jiangxb1987 @srowen @cloud-fan @HyukjinKwon

Could you please authorize my PR for testing?

Thanks
Darek

**Comment by HyukjinKwon** [comment]
ok to test

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaRDD.scala [hunk 2]
```diff
@@ -211,8 +211,8 @@ private[spark] class KafkaRDD[K, V](
     var requestOffset = part.fromOffset
 
     def closeIfNeeded(): Unit = {
-      if (!useConsumerCache && consumer != null) {
-        consumer.close
+      if (consumer != null) {
+          consumer.close()
```
Comment (HyukjinKwon) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaRDD.scala]: I think this should be double spaced

**Review by HyukjinKwon — COMMENTED** [review]
Does this look good to you @koeninger?

**Comment by cloud-fan** [comment]
cc @zsxwing

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaRDD.scala (hunk 2 above)
Comment (daroo) [reply external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaRDD.scala]: formatting changed

**Comment by koeninger** [comment]
Seems reasonable to me but you should probably ask zsxwing if it fits in
with plans for the structured streaming kafka code.

**Comment by daroo** [comment]
I don't know what are the plans for the structured streaming which @zsxwing has in mind, but before I created the PR I actually had seen how this problem is currently solved in kafka-0-10-sql module and I didn't like it. My main concern is that in certain cases the cache size my grow well beyond configured spark.sql.kafkaConsumerCache.capacity. That’s why I’ve chosen to do it in a bit different way.

**Comment by daroo** [comment]
@HyukjinKwon
@zsxwing 

Any chance to go ahead with this?

**Comment by HyukjinKwon** [comment]
I think we should wait for @zsxwing's feedback.

**Comment by daroo** [comment]
Hey @zsxwing

Could you please have a look at my PR? It's been a month almost since I opened it. @koeninger said that it looks good to him, but everyone else is waiting for your feedback.

Thanks a lot,
Darek

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by daroo** [comment]
Hi @HyukjinKwon

It seems that @zsxwing is not really interested in this PR. Should I cancel it and close https://issues.apache.org/jira/browse/SPARK-22562 ?

BTW. Why AppVeyor CI build takes so looooong time? The build has failed due to reaching max allowed time (1,5h) :-(

**Comment by cloud-fan** [comment]
cc @marmbrus @zsxwing

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
@daroo, mind reopening this if you have some time to update?

**Comment by daroo** [comment]
sure :-)

**Comment by daroo** [comment]
@HyukjinKwon I've looked at the latest changes in the code and I think this problem shouldn't happen (at least in practice) anymore. So the JIRA can be closed.
One minor issue I've noticed is that the cache size can keep growing beyond configured limit in certain cases.

**Comment by HyukjinKwon** [comment]
@daroo thanks for checking it.

**Pull request closed without merging**